### PR TITLE
Added custom annotation class, AnnotationProvider and added hierarchy information to member shapes

### PR DESCRIPTION
**Pull request by nowraiz**
## Description

Added custom annotation class (Traceable) for source code annotations, added hierarchy information to nested shapes and implemented a generic AnnotationProvider (which might change in the future).

### CODE REVIEW

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/MemberModel.java [hunk 1]
```diff
@@ -124,6 +126,19 @@ public void setC2jName(String c2jName) {
         this.c2jName = c2jName;
     }
 
+    public void setParentName(String parentName) {
```
Comment (senguptaaritra) [inline codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/MemberModel.java:129]: Are they not using lombok annotations for getters and setters?

File: codegen/src/main/java/software/amazon/awssdk/codegen/tracing/AnnotationProvider.java [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+import com.squareup.javapoet.AnnotationSpec;
+import software.amazon.awssdk.annotations.Traceable;
+
+public class AnnotationProvider {
+    /*
+     This class provides the Annotations for different entities in the code depending upon their type
```
Comment (senguptaaritra) [inline codegen/src/main/java/software/amazon/awssdk/codegen/tracing/AnnotationProvider.java:11]: Typical convention at least DG/DS is to put comments right above the class declaration.

File: codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/MemberModel.java (hunk 1 above)
Comment (nowraiz) [reply codegen/src/main/java/software/amazon/awssdk/codegen/model/intermediate/MemberModel.java:129]: Not really, I don't see any lombok annotations in the code that's why I decided to write the getters and setter separately.

File: codegen/src/main/java/software/amazon/awssdk/codegen/tracing/AnnotationProvider.java (hunk 2 above)
Comment (nowraiz) [reply codegen/src/main/java/software/amazon/awssdk/codegen/tracing/AnnotationProvider.java:11]: Yes. I will change that and follow the convention in the future.

**Pull request merged**